core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1501,13 +1501,16 @@ class KafkaApis(val requestChannel: RequestChannel,
       return
     }
 
+    def updateErrors(producerId: Long, currentErrors: util.HashMap[TopicPartition, Errors]): Unit = {
+      val previousErrors = errors.putIfAbsent(producerId, currentErrors)
+      if (previousErrors != null)
+        previousErrors synchronized previousErrors.putAll(currentErrors)
```

Comment:
This is not quite right. It introduces a happens before during writes, but not when we read from the HashMap. It seems safer to use `Collections.synchronizedMap` or `ConcurrentHashMap` for the inner map. I'd do with the latter. Any reason not to do that?